src/types/utils.ts

```diff
@@ -257,3 +245,152 @@ export function typeNameToTypeNode(astT: TypeName): TypeNode {
+
+    let size: number | undefined;
+
+    for (let n = 8; n <= 32; n += 8) {
+        if (length < 2 ** n) {
+            size = n;
+
+            break;
+        }
+    }
+
+    assert(
+        size !== undefined,
+        "Unable to detect enum type size - member count exceeds 2 ** 32",
+        decl
+    );
+
+    return new IntType(size, false);
+}
+
+/**
+ * Converts `type` to value type. Following rules are applied:
+ * 1. Contract definitions turned to address.
+ * 2. Enum definitions turned to uint of minimal fitting size.
+ * 3. Struct definitions turned to tuples.
+ *   3.1. Mappings are completely ommited.
+ *   3.2. Arrays of root structs are ommited, while preserved for nested structs.
+ * 4. Data locations are skipped.
+ */
+function toValueType(type: TypeNode, depth = 0): TypeNode {
+    if (type instanceof PointerType) {
+        return toValueType(type.to, depth);
+    }
+
+    if (type instanceof UserDefinedType) {
+        if (type.definition instanceof ContractDefinition) {
+            return new AddressType(false);
+        }
+
+        if (type.definition instanceof EnumDefinition) {
+            return enumToIntType(type.definition);
+        }
+
+        if (type.definition instanceof StructDefinition) {
+            const elements: TypeNode[] = [];
+
+            for (const member of type.definition.vMembers) {
+                const memberT = member.vType;
+
+                assert(memberT !== undefined, "Unexpected untyped struct member", type.definition);
+
+                if (memberT instanceof Mapping) {
+                    continue;
+                }
+
+                if (memberT instanceof ArrayTypeName && depth === 0) {
+                    continue;
+                }
+
+                elements.push(
+                    toValueType(
+                        typeNameToSpecializedTypeNode(memberT, DataLocation.Memory),
+                        depth + 1
+                    )
+                );
+            }
+
+            return new TupleType(elements);
+        }
+    }
+
+    return type;
+}
+
+/**
+ * For given variable declaration `decl` computes accessor arguments and return types.
+ * Usage:
+ * ```
+ * const [argTs, retT] = getterArgsAndReturn(v);
+ * ```
+ * Where `argsTs` is an array of computed argument types and `retT` is computed return type.
+ */
+export function getterArgsAndReturn(decl: VariableDeclaration): [TypeNode[], TypeNode] {
+    const argTypes: TypeNode[] = [];
+
+    let type = decl.vType;
+
+    assert(
+        type !== undefined,
+        "Called getterArgsAndReturn() on variable declaration without type",
+        decl
+    );
+
+    while (true) {
+        if (type instanceof ArrayTypeName) {
+            argTypes.push(new IntType(256, false));
+
+            type = type.vBaseType;
+        } else if (type instanceof Mapping) {
+            argTypes.push(
+                toValueType(typeNameToSpecializedTypeNode(type.vKeyType, DataLocation.Memory))
```

Comment:
Why are we calling `toValueType` here? I would think that if we are computing the type for a getter for var `mapping(string => uint)` the argument would be `string memory` not `string`?

Reply:
Due to `Enum`s possible as mapping keys.

Reply:
Regarding `string memory`, see https://github.com/ConsenSys/solc-typed-ast/commit/83cda55cac6638e4dd593bf56869eb33b22584d4. Allowed non-definition types to be pointers. Also pointers are reconstructed to point to `memory` data location.